crates/interpreter/src/instructions/memory.rs

```diff
@@ -35,3 +42,31 @@ pub fn msize(interpreter: &mut Interpreter, _host: &mut dyn Host) {
     gas!(interpreter, gas::BASE);
     push!(interpreter, U256::from(interpreter.memory.effective_len()));
 }
+// From EIP-5656 MCOPY
+pub fn mcopy<SPEC: Spec>(interpreter: &mut Interpreter, _host: &mut dyn Host) {
+    // check that Cancun spec is enabled
+    check!(interpreter, SPEC::enabled(CANCUN));
+    // get src and dest and length from stack
+    pop!(interpreter, dest, src, len);
+
+    // into usize or fail
+    let len = as_usize_or_fail!(interpreter, len, InstructionResult::InvalidOperandOOG);
+    let dest = as_usize_or_fail!(interpreter, dest, InstructionResult::InvalidOperandOOG);
+    let src = as_usize_or_fail!(interpreter, src, InstructionResult::InvalidOperandOOG);
+    // deduce gas
+    gas_or_fail!(interpreter, gas::verylowcopy_cost(len as u64));
+
+    // check if len is 0
+    if len == 0 {
+        return;
+    }
+    // temporary value for memory resize
+    let mut resize = dest;
+    if src + len > dest {
+        resize = src + len;
+    }
```

Comment:
There are two problems here. The first consensus is that `dest+len` should be compared (not just `dest`) and the second is that `+` can overflow.

```suggestion
    // memory resize
    let resize = max(dest, len).saturating_add(len);
```